### ISSUE TITLE
No Operation named [input_tensor] in the Graph

### ISSUE THREAD

**Issue by phydesmith** [tensorflow/java#312]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- I am using code based on [https://github.com/davpapp/PowerMiner/blob/master/src/ObjectDetector.java](url)
- Windows 10 64-bit
- TensorFlow installed from (source or binary): tensorflow-core-platform snapshot from sonatype repo (maven)
- TensorFlow version (use command below): tensorflow-core-platform 0.4.0 snapshot -> TF 2.4 (?)
- Python version: 2.8
- GPU model and memory: NVidia 1080 GTX

**Describe the current behavior**
I have tried two things: 
```
Map<String, Tensor> output = smb.call(input);
```

This gives me the same output no matter the image, and I've looked at the data in the byte buffers and it's different. So I moved to try and do what I found on the above source (which I think is pulled from an old example?)

```
outputsList = smb.session()
                            .runner()
                            .feed("[input_tensor]", input)
                            .fetch("detection_scores")
                            .fetch("detection_classes")
                            .fetch("detection_boxes")
                            .run();
            input.close();
```

This gives me an error saying `Caused by: java.lang.IllegalArgumentException: No Operation named [input_tensor] in the Graph`

I have run the saved_model_cli command and gotten the following output:

```
$ saved_model_cli.exe show --dir saved_model/ --tag_set serve --signature_def serving_default
2021-05-01 13:43:21.409559: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library cudart64_110.dll
The given SavedModel SignatureDef contains the following input(s):
  inputs['input_tensor'] tensor_info:
      dtype: DT_UINT8
      shape: (1, -1, -1, 3)
      name: serving_default_input_tensor:0
```

**Describe the expected behavior**
I expected the call function to just kind of work out of the box and give me the expected output results, I wouldn't be surprised if I have something messed up with this. Though I did test the model (on test images in a google colab python notebook) before it was exported, and it was giving me sane/different outputs based on image inputs.

The second way I was expecting to actually work, consider the saved_model_cli says 'input_tensor' is what it's supposed to be. However, I don't know if those examples are wrong and there should be some other way for me to do the inference? I've had a hard time finding examples.

**Code to reproduce the issue**
This is the class I am using -
[https://gist.github.com/phydesmith/96bf57c2a75e01e698f8e0ebc2c8194c](https://gist.github.com/phydesmith/96bf57c2a75e01e698f8e0ebc2c8194c)

**Comment by Craigacp** [comment]
The first way you tried should have worked, I'm surprised that it doesn't. Did you look at this [object detection example](https://github.com/tensorflow/java-models/tree/master/tensorflow-examples/src/main/java/org/tensorflow/model/examples/cnn/fastrcnn). It's still being tidied up, but it shows loading in a saved model bundle for object detection and then getting the annotated model back out.

In general using a 4d input byte array isn't very fast, but I think BufferedImage stores the data in a different order to the way that TF wants it. We should probably add a helper that does the necessary reshaping.

**Comment by DavidHuebner** [comment]
Regarding the second example, try using `serving_default_input_tensor:0` as the name of the input tensor instead of `input_tensor`.

**Comment by karllessard** [comment]
For the second example, you've feed `[input_tensor]`, probably want to feed `input_tensor` instead.

For the first, is the image preprocessing done correctly, i.e. the image is rescale and pixel values are normalized between 0 and 1 (that's pretty usual)?

**Comment by phydesmith** [comment]
So @DavidHuebner 's solution for the input is getting me somewhere, however it then gives me issues with the outputs when I call `fetch()`. With a similar message of  `No Operation named [serving_default_detection_scores:0] in the Graph` or `No Operation named [serving_default_detection_scores] in the Graph`. 

Here is the output of me calling `this.smb.function("serving_default").signature().getInputs()` and `this.smb.function("serving_default").signature().getOutputs()`.

```
{input_tensor=org.tensorflow.Signature$TensorDescription@4f885aaf}

{raw_detection_boxes=org.tensorflow.Signature$TensorDescription@238cbc, detection_classes=org.tensorflow.Signature$TensorDescription@5792a558, detection_boxes=org.tensorflow.Signature$TensorDescription@6276f97f, num_detections=org.t
ensorflow.Signature$TensorDescription@6f268abc, raw_detection_scores=org.tensorflow.Signature$TensorDescription@73fcad88, detection_scores=org.tensorflow.Signature$TensorDescription@31066eca, detection_anchor_indices=org.tensorflow.
Signature$TensorDescription@646ec494, detection_multiclass_scores=org.tensorflow.Signature$TensorDescription@3f46d1cf}
```

**Comment by karllessard** [comment]
Ok I got it wrong at first. So this model has probably been trained on TF 2.x, which mangles the name of the input/output tensors to be fed to the session.  The `ConcreteFunction` class was initially design to handle this case, so by calling `call`, it will take care of feeding and fetching the right (mangled) names while at the user level, you only need to take care of the signature input/output names (BTW in your previous post, you've print out the output of `getInputs()` but should be `inputNames()` to see the names to set in your input map passed to call the function.

So you did it the right way to begin with by invoking `smb.call(input)` and you shouldn't need to play directly with the session. Now is the issue about how to feed a session to obtain the same results than with the function, or is it that you have doubts about the accuracy of these results? For the latter, if you can provide us more details on how you pass the inputs to the function and read back its output, that could be useful.

**Comment by phydesmith** [comment]
Ok, yes this model was created from the TF 2 Object Detection API. I went back to using the `smb.call(inputMap)` and it gives the same results every time. See [https://gist.github.com/phydesmith/3ee170f806757f1b441ad7be2cd99cc1](url) for the code. Please forgive the hackiness of it at the moment, I'm trying to just get something working and then go back and optimize it to run better.

Basically, I'm doing no image preprocessing aside from converting the `BufferedImage` to a Tensor which I can feed into the model. You can see on the gist I linked that I basically copied the example code and just tried to update it to work with how I thought it should with the current state of the API. 

I can clearly see the byte values differ based on different images I take a screen shot of when I look into the tensor data before I feed it into the model. It seems like every time I use the `call()` function however, it just returns the same outputMap. These returned values are even the same across different runs of my program.

**Comment by pyrator** [comment]
I did a quick check of your processimage code and compared it to the output from a DecodePng and it does appear that you are mixing up your RGB values.
Here's my gist for creating and comparing a 3 * 3 image
https://gist.github.com/pyrator/b7433522b083c073a3986aea7969e9ac

**Comment by phydesmith** [comment]
@pyrator  I have tried incorporating your changes and I'm still seeing the same issues. I have tried it with a different but similarly trained model and it's giving me the same scores and bounding boxes from different image inputs.

**Comment by pyrator** [comment]
I've double checked my code and added and EncodePng and saved the resulting PNG and it looks the same to me. TBH you're probably creating a rod for your own back doing it this way. I can understand you might be wanting to understand the inner workings of tensors, and how they relate to images, but there are methods to Decode images available - see my gist above. Also check out the example at 
https://github.com/tensorflow/java-models/tree/master/tensorflow-examples/src/main/java/org/tensorflow/model/examples/cnn/fastrcnn

**Comment by karllessard** [comment]
While @pyrator is right that using one of the `Decode*` op is safer that manipulating the image directly, you can still convert directly a `BufferedImage` to a `Tensor` by manipulating its pixels direclty, I have an application running in production for years doing this and it works fine.

Please @phydesmith , take a look at this [snippet](https://discuss.tensorflow.org/t/creating-float-tensors-from-bufferedimage-in-java-kotlin/1865) I've added a few month ago. Your gist is no longer available but I suspect that not only you might have been reverting your RGB channels (`BufferedImage` are BGR) but you might have forgot the normalization step where you want each value of the tensor to be a float between 0 and 1. 

The snippet above shows how Keras does it by default but there are other normalization modes that could have been enabled during training, if you still encounter issues double check that default normalization (caffe) has been used for training.

**Comment by phydesmith** [comment]
I'm going to close this issue and move to the thread Karl linked for further discussion as my main problem was solved.
